src/java.desktop/macosx/native/libosxapp/JNIUtilities.h

```diff
@@ -0,0 +1,161 @@
+/********        LOGGING SUPPORT    *********/
+
+#define LOG_NULL(dst_var, name) \
+   if (dst_var == NULL) { \
+       NSLog(@"Bad JNI lookup %s\n", name); \
+    }
+
+/********        GET CLASS SUPPORT    *********/
+
+#define GET_CLASS(dst_var, cls) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->FindClass(env, cls); \
+         if (dst_var != NULL) dst_var = (*env)->NewGlobalRef(env, dst_var); \
+     } \
+     LOG_NULL(dst_var, cls); \
+     CHECK_NULL(dst_var);
+
+#define DECLARE_CLASS(dst_var, cls) \
+    static jclass dst_var = NULL; \
+    GET_CLASS(dst_var, cls);
+
+#define GET_CLASS_RETURN(dst_var, cls, ret) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->FindClass(env, cls); \
+         if (dst_var != NULL) dst_var = (*env)->NewGlobalRef(env, dst_var); \
+     } \
+     LOG_NULL(dst_var, cls); \
+     CHECK_NULL_RETURN(dst_var, ret);
+
+#define DECLARE_CLASS_RETURN(dst_var, cls, ret) \
+    static jclass dst_var = NULL; \
+    GET_CLASS_RETURN(dst_var, cls, ret);
+
+
+/********        GET METHOD SUPPORT    *********/
+
+#define GET_METHOD(dst_var, cls, name, signature) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetMethodID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL(dst_var);
+
+#define DECLARE_METHOD(dst_var, cls, name, signature) \
+     static jmethodID dst_var = NULL; \
+     GET_METHOD(dst_var, cls, name, signature);
+
+#define GET_METHOD_RETURN(dst_var, cls, name, signature, ret) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetMethodID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL_RETURN(dst_var, ret);
+
+#define DECLARE_METHOD_RETURN(dst_var, cls, name, signature, ret) \
+     static jmethodID dst_var = NULL; \
+     GET_METHOD_RETURN(dst_var, cls, name, signature, ret);
+
+#define GET_STATIC_METHOD(dst_var, cls, name, signature) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetStaticMethodID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL(dst_var);
+
+#define DECLARE_STATIC_METHOD(dst_var, cls, name, signature) \
+     static jmethodID dst_var = NULL; \
+     GET_STATIC_METHOD(dst_var, cls, name, signature);
+
+#define GET_STATIC_METHOD_RETURN(dst_var, cls, name, signature, ret) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetStaticMethodID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL_RETURN(dst_var, ret);
+
+#define DECLARE_STATIC_METHOD_RETURN(dst_var, cls, name, signature, ret) \
+     static jmethodID dst_var = NULL; \
+     GET_STATIC_METHOD_RETURN(dst_var, cls, name, signature, ret);
+
+/********        GET FIELD SUPPORT    *********/
+
+
+#define GET_FIELD(dst_var, cls, name, signature) \
+     if (dst_var == NULL) { \
+         dst_var = (*env)->GetFieldID(env, cls, name, signature); \
+     } \
+     LOG_NULL(dst_var, name); \
+     CHECK_NULL(dst_var);
```

Comment:
If some of these macros will be used on appkit thread, then the check like "CHECK_NULL" will not work,  because the code will never return to the java, and the failure will be actually ignored.

Reply:
If we get NULL back then there's a progamming bug. LOG_NULL will show it.
CHECK_NULL is just as much to satisfy JNI as anything.
If we are on a Java thread then it will still add some value so I don't see the point in removing it.
Nor do I think I'd like to use NSException instead because that affects the flow of control.
Yes, this may be different than before (probably more relevant to the other comment) but I think better may be a better approach overall. If we ever see an issue here that is likely to occur due to anything other than a bad internal inconsistency due to JDK changes then I would prefer to follow the code path and fix it up.

Reply:
I mean all macro, not just CHECK_NULL. You just create a JNF library 1.5 which for some reason uses macro instead of objective-c functions, uses logging instead of exceptions, and changes a huge number of files. Does it make the codebase better?
 - I think we should ask Apple to contribute the JNF library or merge it as a third-party library, then cut all unused parts of it
 - Or we should create JNF library 2.0, and the whole fix will be looking like a replacement of three letters in each function, like JNFCallVoidMethod - > JNUCallVoidMethod etc.

Reply:
I have to disagree. (1) is a non-starter and (2) is un-needed and I have no desire to copy the patterns in JNF
which I don't think are needed. This will end up simpler.